storage/replica.go

```diff
@@ -911,16 +1061,61 @@ func (r *Replica) applyRaftCommandInBatch(ctx context.Context, index uint64, ori
 		}
 	}
 
-	// Execute the command.
-	reply, intents, rErr := r.executeCmd(batch, ms, args)
+	bReply := &proto.BatchResponse{}
+	bReply.Timestamp = bArgs.Timestamp
+	err, cache := func() (error, bool) {
+		for _, union := range bArgs.Requests {
+			args := union.GetValue().(proto.Request)
+
+			if lease := r.getLease(); args.Method() != proto.LeaderLease &&
```

Comment:
I would keep the lease check out of the loop - all commands in the batch should be covered (or not) by the same lease. The `LeaderLease` method should be one of those that is only allowed in a batch by itself.